accepted/2021/statics-in-interfaces/README.md

```diff
@@ -8,6 +8,793 @@ C# is looking at enabling static abstract members in interfaces (https://github.
+
+```csharp
+namespace System
+{
+    public interface INumber<TSelf>
+        : IAddable<TSelf, TSelf>,
+          IComparableOperators<TSelf, TSelf>,   // implies IEquatableOperators<TSelf, TSelf>
+          IDecrementable<TSelf>,
+          IDivisable<TSelf, TSelf>,
+          IIncrementable<TSelf>,
+          IMultipliable<TSelf, TSelf>,
+          IRemainder<TSelf, TSelf>,
+          ISpanFormattable,                     // implies IFormattable
+          ISpanParseable<TSelf>,                // implies IParseable<TSelf>
+          ISubtractable<TSelf, TSelf>
+        where TSelf : INumber<TSelf>
+    {
+        // For the Create methods, there is some concern over users implementing them. It is not necessarily trivial to take an arbitrary TOther
+        // and convert it into an arbitrary TSelf. If you know the type of TOther, you may be able to optimize in a few ways. Otherwise, you
+        // may have to fail and throw in the worst case.
+
+        static abstract TSelf Create<TOther>(TOther value)
+            where TOther : INumber<TOther>;
+
+        static abstract TSelf? CreateExactly<TOther>(TOther value)
+            where TOther : INumber<TOther>;
+
+        static abstract TSelf CreateSaturating<TOther>(TOther value)
+            where TOther : INumber<TOther>;
+
+        static abstract TSelf CreateTruncating<TOther>(TOther value)
+            where TOther : INumber<TOther>;
+
+        // There is an open question on whether properties like IsSigned, IsBinary, IsFixedWidth, Base/Radix, and others are beneficial
+        // We could expose them for trivial checks or users could be required to check for the corresponding correct interfaces
+
+        // MinValue and MaxValue are more general purpose than just "numbers", they should probably live on some common base type
+        // Likewise, the concept is a bad fit for types like BigInteger which aren't fixed sized
+        // There may also be concern around the primitive types which already define these names as constants
+
+        static abstract TSelf MinValue { get; }
+
+        static abstract TSelf MaxValue { get; }
+
+        static abstract TSelf One { get; }
+
+        static abstract TSelf Zero { get; }
+
+        // Abs mirrors Math.Abs and returns the same type. This can fail for MinValue of signed integer types
+        // Swift has an associated type that can be used here, which would require an additional type parameter in .NET
+        // However that would hinder the reusability of these interfaces in constraints
+
+        static abstract TSelf Abs(TSelf value);
+
+        static abstract TSelf Clamp(TSelf value, TSelf min, TSelf max);
+
+        static abstract (TSelf Quotient, TSelf Remainder) DivRem(TSelf left, TSelf right);
+
+        static abstract TSelf Max(TSelf x, TSelf y);
+
+        static abstract TSelf Min(TSelf x, TSelf y);
```

Comment:
These methods should be part of `IComparableOperators`.

As discussed, complex numbers or matrix algebras cannot define total orders and as such `Min` and `Max` operators cannot be reasonably defined.

Reply:
Moving them to `IComparable` wouldn't help with `Complex` or `Matrix` like types, given that `INumber` implements `IComparable`.

It would, however, impact types like `Guid`, `DateTime`, `TimeSpan` and others that don't define `Min`, `Max`, or `Clamp` but would implement `IComparable`. I don't see an issue here right now, but thought I'd call it out.

---

On a meta point, I think the interesting split between `Complex`, `Vector`, `Matrix`, and `INumber` is that the latter really defines something that can be interpreted as a scalar value. A `Vector` is composed of two or more scalars and a `Matrix` is composed of two or more vectors. In that sense, a `Complex` number is most like a vector in that it has two scalar parts (a real part `a` and an imaginary part `bi`).

How `Complex` composes here needs some more thought and prototyping to determine how it can be exposed, particularly given I expect most users will largely care about this from the "scalar" (single value) concept for `INumber` in typical use-cases.

Likewise, while `Matrix` and `Vector` are often not `IComparable` in the traditional sense, they do often support these operations returning `TSelf` for an "element-wise" comparison, which is often used as a mask for downstream operations and sometimes as an "do all elements follow this comparison" (as is the case for `Vector<T>`).

Reply:
> Moving them to IComparable wouldn't help with Complex or Matrix like types, given that INumber implements IComparable.

Right, I was suggesting that perhaps `INumber` should not inherit the comparison interface. 

> It would, however, impact types like Guid, DateTime, TimeSpan and others that don't define Min, Max, or Clamp but would implement IComparable. I don't see an issue here right now, but thought I'd call it out.

It is possible to express `Min`, `Max` and `Clamp` in terms of comparison, so perhaps we could add those as DIMs in the comparison interface.

> On a meta point, I think the interesting split between Complex, Vector, Matrix, and INumber is that the latter really defines something that can be interpreted as a scalar value. A Vector is composed of two or more scalars and a Matrix is composed of two or more vectors. In that sense, a Complex number is most like a vector in that it has two scalar parts (a real part a and an imaginary part bi).

From a memory representation perspective that's a pretty accurate distinction, in fact it's fairly common to define complex numbers using [matrices](https://en.wikipedia.org/wiki/Complex_number#Matrix_representation_of_complex_numbers). From an abstraction perspective however complex numbers are numbers, and an abstraction called `INumber` should reasonably support those as well.

> Likewise, while Matrix and Vector are often not IComparable in the traditional sense, they do often support these operations returning TSelf for an "element-wise" comparison, which is often used as a mask for downstream operations and sometimes as an "do all elements follow this comparison" (as is the case for Vector<T>).

I suspect though that such a comparison would not be a total order and therefore cannot be used meaningfully for things like building a search tree.

Reply:
> Right, I was suggesting that perhaps INumber should not inherit the comparison interface.

That will hinder what I've currently tracked as the "most common scenario", which is abstracting over the 13 primitive types (`sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `nint`, `nuint`, `long`, `ulong` as well as `float`, `double`, and `half`). The "typical" constraint would then become `where T : INumber<T>, IComparableOperators<T, T>`. Types like `char` and `decimal` are also sometimes included here, as well as types like `BigInteger` and a conceptual future `BigFloat` or types like `Int128`, `UInt128`, `decimal32`, `decimal64`, `decimal128` (the IEEE decimal types) and `Quad` (the IEEE `binary128`). 

Types like `Complex`, while conceptually "numbers", really have a different "interface" and handling when thinking from an OOP programming algorithm. I would think the same applies to things like `Vector` or `Matrix` where even though they support many of the operations, you want to consider and handle them differently and would likely have a different overload to do so.

We could have something like `INumber` vs `IScalarNumber`, but there is cost (including runtime startup cost) to every one of the interfaces here so we need to be careful about making this "too expansive" while still allowing the correct user scenarios.

Reply:
> We could have something like INumber vs IScalarNumber, but there is cost (including runtime startup cost) to every one of the interfaces here so we need to be careful about making this "too expansive" while still allowing the correct user scenarios.

We could perhaps strike a balance by doing away some of the other intermediate interfaces (e.g. `IAddable`). Really though I find this to be a naming issue, we could consider renaming `INumber` to something like `IScalarNumber` or `IPrimitiveNumber` to avoid confusion.

Reply:
> Really though I find this to be a naming issue

Right, this is roughly based on what Swift calls it (`Numeric` protocol) where they likewise imply similar requirements/guarantees around Comparisons, Equality, etc.
We could clarify by calling it something else or by having an intermediate step between "numbers" and "scalar numbers".

> We could perhaps strike a balance by doing away some of the other intermediate interfaces (e.g. IAddable).

I think that will be too limiting and I'd rather have a strong base here that allows other libraries to define their own contracts as appropriate. Otherwise, we basically end up with `IComparableOperators`, `IEquatableOperators`, and `INumber` (plus potentially something like `IVector`) and so DateTime and others won't really be supported here.

Reply:
Do we have a use case for DateTime and friends benefiting from generic arithmetic?

Reply:
Not strictly speaking. They are interesting as they are types where a user may want to operate over the lower base interfaces (e.g. IAddable) rather than the higher abstractions (e.g. INumber).

You could imagine a `Sum<T> where T : IAddable<T, T>` method or similar being a valid use-case here.

Reply:
The same applies for the general comparison, equality, min, max, and similar functionality that could be defined for types like DateTime, TimeSpan, Guid, etc.

Reply:
> You could imagine a Sum<T> where T : IAddable<T, T> method or similar being a valid use-case here.

Right, so effectively defining reduce over a monoid-like abstraction. In practice I find that the definition of addition greatly varies depending on the application, and it might often not coincide with what the type actually implements (e.g. reduce using multiplication or some other binary operation). It is usually simpler to just use LINQ's `Aggregate`.

> The same applies for the general comparison, equality, min, max, and similar functionality that could be defined for types like DateTime, TimeSpan, Guid, etc.

There's certainly value in doing that. I'm just not convinced that addition is as useful for those particular types.

Reply:
> It is usually simpler to just use LINQ's Aggregate.

Many of these concepts came up as a question of "if we were to implement LINQ today, would they take `T where T : IAddable<T>`"

I imagine most algorithms don't care that it is a monoid, they just care that you can do:
```csharp
T sum = default; 

foreach (var value in values)
{
    sum += value;
}

return sum;
```

This like commutativity, associativity, distributivity and others (particularly in programming) largely only come up in the context of optimizations (like constant folding). They almost never come up otherwise (outside maybe simplification, which would be an analyzer, not something C# does itself).

Reply:
I wasn't really making a point about monoid laws, and you're right that most simple reduce operations don't depend on them holding. I'm merely questioning the usefulness of `T where T : IAddable<T>` methods in the grander scheme of things.

Reply:
It's definitely possible that in practice they won't be used and that's something we'll need to determine before shipping.

At this time, and based on other languages/frameworks, they seem like good/practical building blocks that we'll want to expose.

Will definitely keep this in mind as the community outreach begins.